fastroute caching support

pine3ree opened the pull request:
@weierophinney 
I haven't added tests yet, but i tested the cached routes generation with a small blog app. 
You are welcome to review the code when you have time (when jumping from one zf3 component to another :-) ) 
I have an addon feature that instead of loading the cache inside the constructor delegates this operation to an extra `initialize()` method (plus an `$initialized` control property), called as first step inside `match()` and `generateUri()`.

pine3ree committed:
fastroute caching support

pine3ree committed:
caching disabled by default

pine3ree committed:
remove whitespaces

pine3ree commented:
we would also need to create the instance using a dedicated factory and inject config parameters into the options constructor as explained here https://github.com/zendframework/zend-expressive-fastroute/issues/9, maybe using a specific config key `fastroute`

pine3ree committed:
renamed cosntructor parameters

$options to $config to avoid visual confusion with route options

pine3ree committed:
added factory class

pine3ree committed:
renamed config keys using standard zend snake case

thus avoiding conversion like cache_file => cacheFile in the factory

pine3ree commented:
@weierophinney 
i added a small factory to pass configuration options into the FastRouteRouter constructor.
We also need to change `routes.global.php` configuration file:

``` php
return [
    'dependencies' => [
        //..
        'factories' => [
            //..
            Zend\Expressive\Router\RouterInterface::class => Zend\Expressive\Router\FastRouteRouterFactory::class,
            //..
        ],
    ],

    'router' => [
        'fastroute' => [
            'cache_enabled' => true,
            'cache_file'    => 'data/cache/fastroute.cache',
        ],
    ],

    'routes' => [...],
]
```

pine3ree commented:
@weierophinney 
Hello, do you think the following code could be sufficient and appropriate for testing fastroute caching

``` php
    //...
    public function createCachingRouter(array $config, Route $route)
    {
        $router = new FastRouteRouter(null, null, $config);
        $router->addRoute($route);

        return $router;
    }

    public function createServerRequestProphecy($path, $method = 'GET')
    {
        $uri = $this->prophesize(UriInterface::class);
        $uri->getPath()->willReturn($path);

        $request = $this->prophesize(ServerRequestInterface::class);
        $request->getUri()->will(function () use ($uri) {
            return $uri->reveal();
        });

        $request->getMethod()->willReturn($method);

        return $request;
    }

    public function testFastRouteCache()
    {
        $cache_file = __DIR__ . '/fastroute.cache';

        $config = [
            'cache_enabled' => true,
            'cache_file'    => $cache_file,
        ];

        $request = $this->createServerRequestProphecy('/foo', 'GET');

        $route = new Route('/foo', 'fooHandler', ['GET'], 'foo');

        $router1 = $this->createCachingRouter($config, $route);
        $router1->match($request->reveal());

        $this->assertTrue(is_file($cache_file));

        $mtime1 = filemtime($cache_file);
        $cache1 = file_get_contents($cache_file);

        usleep(1 + 1000 * 1000);

        $router2 = $this->createCachingRouter($config, $route);

        $result = $router2->match($request->reveal());

        $this->assertTrue(is_file($cache_file));

        $mtime2 = filemtime($cache_file);
        $cache2 = file_get_contents($cache_file);

        $this->assertEquals($mtime1, $mtime2);
        $this->assertEquals($cache1, $cache2);

        $this->assertInstanceOf(RouteResult::class, $result);
        $this->assertTrue($result->isSuccess());
        $this->assertEquals('foo', $result->getMatchedRouteName());
        $this->assertEquals('fooHandler', $result->getMatchedMiddleware());

        unlink($cache_file);
    }
}
```

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -114,12 +159,14 @@ public function match(Request $request)
         $result     = $dispatcher->dispatch($method, $path);
 
-        if ($result[0] != Dispatcher::FOUND) {
+        if ($result[0] !== Dispatcher::FOUND) {
```
Change possibly unrelated

Ocramius replied:
Note: change is good, just need to make sure it doesn't break anything. There should be a test in place justifying this change too

pine3ree replied:
@Ocramius 
Yes, it's unrelated. $result[0] from FastRoute dispatcher dispatch() call is always either `FastRoute\Dispatcher::NOT_FOUND` (int 0)
or
`FastRoute\Dispatcher::FOUND` (int 1)
or
`FastRoute\Dispatcher::METHOD_NOT_ALLOWED` (int 2)

ref: https://github.com/nikic/FastRoute/blob/8ea928195fa9b907f0d6e48312d323c1a13cc2af/src/Dispatcher.php

Ocramius replied:
Alright. I don't like that the interface doesn't enforce them, but I suppose there's no way to do that without some massive overhead in there...

pine3ree replied:
@Ocramius 
I believe I asked the author for a Result object a long time ago, but if I remember correctly he wanted to keep that part of the code as simpler as possible, without interfaces, etc etc...

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -80,15 +109,31 @@ class FastRouteRouter implements RouterInterface
         $this->router = $router;
         $this->dispatcherCallback = $dispatcherFactory;
+
+        if (isset($config['cache_enabled'])) {
```
These 3 conditionals should be moved to a private method, IMO

pine3ree replied:
@Ocramius 
done here: https://github.com/zendframework/zend-expressive-fastroute/pull/16/commits/b51205518de3aae08a51b27cf3fe4c6925f22a4f

I am not sure about the call to loadDispatchData; maybe we could keep only the reading of $config keys in there (for other future extra parameters) and move back that call into the constructor after the loadConfig call. Which option do you think best?

kind regards

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -80,15 +109,31 @@ class FastRouteRouter implements RouterInterface
         $this->dispatcherCallback = $dispatcherFactory;
+
+        if (isset($config['cache_enabled'])) {
+            $this->cacheEnabled = (bool) $config['cache_enabled'];
```
Please use class constants, not just magic strings. We had enough typos due to these in the past

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -80,15 +109,31 @@ class FastRouteRouter implements RouterInterface
+            $this->cacheEnabled = (bool) $config['cache_enabled'];
+        }
+
+        if (isset($config['cache_file'])) {
```
Please use class constants, not just magic strings. We had enough typos due to these in the past

pine3ree replied:
@Ocramius 
Hello Marco. Why class constants?
I meant those keys to be used (and related values loaded from) in configuration files, same way it's done for instance in

https://github.com/zendframework/zend-expressive-zendviewrenderer/blob/master/src/HelperPluginManagerFactory.php.

or like `config['middleware_pipeline']`, `config['routes']`, `config['db']`,, `config['db']['username']`, etc etc

If i write:

``` php
if (isset($config[self::CONFIG_CACHE_FILE])) {
```

then I'd have to use it in config files too instead of a simpler `'cache_file'` config key. I've never seen class constants used in config keys in zend-expressive so far except for FQCNs in services/factories of course .

Or I am missing completely your point? :-)

kind regards

Ocramius replied:
> Why class constants?

Preventing typos, reducing documentation overhead, reducing hidden knowledge, easing refactoring

> I meant those keys to be used (and related values loaded from) in configuration files, same way it's done for instance in https://github.com/zendframework/zend-expressive-zendviewrenderer/blob/master/src/HelperPluginManagerFactory.php.

If I had reviewed [`HelperPluginManagerFactory.php`](https://github.com/zendframework/zend-expressive-zendviewrenderer/blob/ca34ef052cefb48de5de6c519e61bbaf7870ea97/src/HelperPluginManagerFactory.php) when it was written, I would have raised the same issue during review

> then I'd have to use it in config files too

You can still use string keys in config files, but it would be silly and error-prone.

pine3ree replied:
@Ocramius 
I agree with You on all the reasons (Preventing typos, reducing documentation overhead, reducing hidden knowledge, easing refactoring) + ide type-hinting :-).

But i think that every zend component that loads/uses configuration is not using class constants.
Nevertheless, it's not an issue for me to add them. I just tried to use the same "coding" style I saw in other related packages.

kind regards

Ocramius replied:
Do not really care about the coding style myself: safety first

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -326,6 +381,80 @@ private function injectRoute(Route $route)
+     */
+    private function loadDispatchData()
+    {
+        if (! is_file($this->cacheFile)) {
```
can `is_file` be skipped, and instead use something like a direct `include` statement? That would prevent a stat call here

Ocramius replied:
setting+removing an error handler would be more efficient, for example

pine3ree replied:
@Ocramius 
Hello, added here https://github.com/zendframework/zend-expressive-fastroute/pull/16/commits/6d145097d6f8635c89a28aeda0616afd3ff9813c

btw, is it possible to make th ecs test pass with the anonymous empty-body function in one line (previous commit didn't pass the php-56 wher cs_check is enabled); i have seen it somewhere in other zend components?

kind regards

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -326,6 +381,80 @@ private function injectRoute(Route $route)
+
+        $dispatchData = require $this->cacheFile;
+        if (! is_array($dispatchData)) {
+            throw new Exception\RuntimeException(
```
Calls for custom exception class

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -326,6 +381,80 @@ private function injectRoute(Route $route)
+        $cacheDir = dirname($this->cacheFile);
+
+        if (! is_dir($cacheDir)) {
+            throw new Exception\RuntimeException(
```
Calls for custom exception class

Ocramius commented on src/FastRouteRouter.php:
```diff
@@ -326,6 +381,80 @@ private function injectRoute(Route $route)
+        }
+
+        if (! is_writable($cacheDir)) {
+            throw new Exception\RuntimeException(
```
Calls for custom exception class

pine3ree replied:
@Ocramius 
Hello, do you mean a simple local namespace exception like https://github.com/zendframework/zend-expressive-router/blob/master/src/Exception/RuntimeException.php for instance or a properly-named execption like InvalidCacheDirectoryException ?

kind regards

Ocramius commented on src/FastRouteRouterFactory.php:
```diff
@@ -0,0 +1,42 @@
+ * ]
+ * </code>
+ */
+class FastRouteRouterFactory
```
Is this even needed?

pine3ree replied:
Hi, I believe we should allow users to choose weather to use caching or not. 
Modifying the factory so that it can parse some properly named config keys (the config keys i chose are just an example, it's what I use) and passing them to a modified Router constructor makes it more flexible. Otherwise one would have to use a custom factory in order to pass caching-related config parameters.

Ocramius replied:
Yes, but is the factory needed at all? How would it be used in a standard expressive setup?

pine3ree replied:
@Ocramius 
The factory is needed to inject configuration parameters from the container config to the router constructor (btw initially the constructor had 2 extra params $cacheFile and $chacheEnabled, no need for special config keys inside the class, but afterwards i decided to use one $config as it was done in other components)  Since the default properties are set not to use caching, the factory would be needed only for those who want to use caching, but i don't think always using the factory could have any impact on performance as the router is created only once.

in my view  
- the expressive setup would always have the factory taking care of the router creation.
- the factory would examine the configuration and pass a $config array to the router constructor.
- the router would be using or not using caching based on the configuration params.

pine3ree committed:
use class constants for configuration keys as suggested by @Ocramius

pine3ree committed:
added test case for caching

pine3ree committed:
move configuation related clauses out of the constructor and into a dedicated private method as suggested by Ocramius

pine3ree committed:
move configuation related clauses out of the constructor and into a dedicated private method as suggested by Ocramius (updated doc-block)

pine3ree committed:
skip stat call as suggested by Ocramius

pine3ree committed:
skip stat call as suggested by Ocramius (cs-fix for anonymous error handler)

pine3ree commented:
@Ocramius 
I added a test case: I build 2 cache-enabled routers. Both routers are using the same cache file. Adding a route to the first should be reflected in a positive match() call in the second router for a path/uri matching the very same route. In order to do that the 2nd router is created after calling match from the 1st route, so that dispatch data is cached first.

In my original test i also wanted to check that the 2nd route would not write to the cache file, but since checking with filemtime required calling sleep(1)/usleep(1000001) to have at least 1 sec from 2 possible file modifictions, i removed that part (see https://github.com/zendframework/zend-expressive-fastroute/pull/16#issuecomment-217200399)

kind regards

glen-84 commented:
Maybe it should use PSR-6?

geerteltink commented:
@glen-84 It's probably faster to generate all the routes again than starting a complete cache system for only 1 file, which is probably cached by opcache anyway.

pine3ree commented:
@glen-84 
fastroute cache can be considered more like a simple "resolved" configuration cache and there is only 1 value being cached, an array of dispatch data create by the used fastroute data generator.
It's important to use a cache for fast-route since rebuilding the regex on every requests takes away the benifits of a fewer fast combined pattern matching.

kind regards.

mtymek commented:
How far are we from merging this?

glen-84 commented:
@xtreamwayz,

Well, it's like 4 interfaces and maybe 1 implementation file for a simple storage. Also, these files may already be loaded for other uses. In fact, a cache pool object may already be in memory.

Anyway, it was just a thought. (=

pine3ree commented:
Hello @mtymek,
I implemented most of the requests of the reviewer long time ago (even I still don't think we should use class constants for the 2 options keys passed in the constructor, the $options params can be thought as the options params in zend form elements and in this case its keys are not used internally, just used to set property values). As far a I know, there is still an open question about implementing a custom Exception and maybe add some documentation. btw I have been using this pr in production (and a further extension of it) for quite a while.

kind regards

weierophinney commented:
Tests look good. I'll address the custom exception classes during merge.

If you could do a pull request against zendframework/zend-expressive to update the documentation for this router — to cover the new caching options — that would be fantastic.

weierophinney commented:
Merged to develop for release with 1.3.0; thanks, @pine3ree\!
